freqtrade/freqai/data_drawer.py

```diff
@@ -0,0 +1,589 @@
+            trained_timestamp: int = the last time the coin was trained
+            return_null_array: bool = Follower could not find pair metadata
+        """
+        pair_dict = self.pair_dict.get(pair)
+        data_path_set = self.pair_dict.get(pair, {}).get("data_path", None)
+        return_null_array = False
+
+        if pair_dict:
+            model_filename = pair_dict["model_filename"]
+            trained_timestamp = pair_dict["trained_timestamp"]
+        elif not self.follow_mode:
+            pair_dict = self.pair_dict[pair] = {}
+            model_filename = pair_dict["model_filename"] = ""
+            trained_timestamp = pair_dict["trained_timestamp"] = 0
+            pair_dict["priority"] = len(self.pair_dict)
+
+        if not data_path_set and self.follow_mode:
+            logger.warning(
+                f"Follower could not find current pair {pair} in "
+                f"pair_dictionary at path {self.full_path}, sending null values "
+                "back to strategy."
+            )
+            trained_timestamp = 0
+            model_filename = ''
+            return_null_array = True
+
+        return model_filename, trained_timestamp, return_null_array
+
+    def set_pair_dict_info(self, metadata: dict) -> None:
+        pair_in_dict = self.pair_dict.get(metadata["pair"])
+        if pair_in_dict:
+            return
+        else:
+            self.pair_dict[metadata["pair"]] = {}
+            self.pair_dict[metadata["pair"]]["model_filename"] = ""
+            self.pair_dict[metadata["pair"]]["first"] = True
+            self.pair_dict[metadata["pair"]]["trained_timestamp"] = 0
+            self.pair_dict[metadata["pair"]]["priority"] = len(self.pair_dict)
+            return
+
+    def pair_to_end_of_training_queue(self, pair: str) -> None:
+        # march all pairs up in the queue
+        for p in self.pair_dict:
+            self.pair_dict[p]["priority"] -= 1
+        # send pair to end of queue
+        self.pair_dict[pair]["priority"] = len(self.pair_dict)
+
+    def set_initial_return_values(self, pair: str, dk: FreqaiDataKitchen,
+                                  pred_df: DataFrame, do_preds: npt.ArrayLike) -> None:
+        """
+        Set the initial return values to a persistent dataframe. This avoids needing to repredict on
+        historical candles, and also stores historical predictions despite retrainings (so stored
+        predictions are true predictions, not just inferencing on trained data)
+        """
+        # dynamic df returned to strategy and plotted in frequi
+        mrv_df = self.model_return_values[pair] = pd.DataFrame()
+
+        for label in dk.label_list:
+            mrv_df[label] = pred_df[label]
+            mrv_df[f"{label}_mean"] = dk.data["labels_mean"][label]
+            mrv_df[f"{label}_std"] = dk.data["labels_std"][label]
+
+        if self.freqai_info.get("feature_parameters", {}).get("DI_threshold", 0) > 0:
```

Comment:
I think you can safely assume `feature_parameters` to be present based on the schema definition - and as such, access it via `freqai_info['feature_parameters']`.
For obvious reasons, that's a faster call than `.get()` ... which is unnecessary here (at last in thee first segment.)

obviously also applies below ...